cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mesh
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+
+	"istio.io/operator/pkg/manifest"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	meshInfoVersion "istio.io/pkg/version"
+
+	goversion "github.com/hashicorp/go-version"
+
+	"istio.io/operator/pkg/compare"
+	pkgversion "istio.io/operator/pkg/version"
+	opversion "istio.io/operator/version"
+)
+
+const (
+	// The maximum duration the command will wait until the apply deployment reaches a ready state
+	upgradeWaitSecWhenApply = 300
+	// The time that the command will wait between each check of the upgraded version.
+	upgradeWaitSecCheckVerPerLoop = 10
+	// The maximum number of attempts that the command will check for the upgraded version.
```

Comment:
Is there any reason max attempts can't by just calculated from upgradeWaitSecWhenApply? It's not clear from the comments how it's controlling something different from what is controlled by upgradeWaitSecWhenApply.

Reply:
Let me update the comments here. The 'max attempts' is used for validating the pods are all running on the same version (~2 mins); while upgradeWaitSecWhenApply is used for waiting the deployment is triggered(?), which can be really quick.

Reply:
I added this (waiting for termination of old pods) because I found an issue (https://github.com/istio/istio.io/issues/5125): the old version pods cannot be terminated during the upgrade.